File: 4_Lists/julia/unit_tests.jl
include("ListModule.jl")
using .ListModule: DsList, append!, insert!, remove_at!, clear!

failures = String[]
success_count = 0

function check(test_name::String, f::Function)
    global success_count
    try
        f()
        success_count += 1
    catch e
        push!(failures, "$(test_name) -> $(e)")
    end
end

check("test_append") do
    ds = DsList()
    append!(ds, "a")
    append!(ds, "b")
    append!(ds, "c")
    @assert length(ds) == 3
end

check("test_get_item") do
    ds = DsList()
    append!(ds, "a")
    append!(ds, "b")
    append!(ds, "c")
    @assert ds[2] == "b"
end

check("test_set_item") do
    ds = DsList()
    append!(ds, "a")
    append!(ds, "b")
    append!(ds, "c")
    ds[2] = "d"
    @assert ds[2] == "d"
end

check("test_iter") do
    ds = DsList()
    append!(ds, "a")
    append!(ds, "b")
    count = 0
    for _ in ds
        count += 1
    end
    @assert count == 2
end

check("test_add") do
    ds1 = DsList()
    append!(ds1, "e")
    append!(ds1, "b")

    ds2 = DsList()
    append!(ds2, "f")
    append!(ds2, "f")

    ds3 = ds1 + ds2
    @assert string(ds3) == "['e', 'b', 'f', 'f']"
end

check("test_in_true") do
    ds = DsList()
    append!(ds, "e")
    append!(ds, "b")
    append!(ds, "f")
    append!(ds, "f")
    @assert ("b" in ds) == true
end

check("test_in_false") do
    ds = DsList()
    append!(ds, "e")
    append!(ds, "b")
    append!(ds, "f")
    append!(ds, "f")
    @assert length(ds) == 4
    @assert ("z" in ds) == false
end

check("test_clear") do
    ds = DsList()
    append!(ds, "e")
    append!(ds, "b")
    append!(ds, "f")
    append!(ds, "f")
    clear!(ds)
    @assert length(ds) == 0
end

println()
if !isempty(failures)
    for msg in failures
        println("ERROR: ", msg)
    end
    println("\n$(length(failures)) tests failed")
    println("\n$(success_count) tests passed")
    println()
    exit(1)
end

println("All $(success_count) tests passed")
println()

File: 5_Lists/javascript/unit_tests.js
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import DsList from './ListModule.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const failures = [];
let successCount = 0;

function assert(condition, message) {
  if (!condition) {
    throw new Error(message || 'Assertion failed');
  }
}

function testAppend() {
  const dsList = new DsList();
  dsList.append('a');
  dsList.append('b');
  dsList.append('c');
  assert(dsList.length === 3, 'Expected length 3 after append');
}

function testGetItem() {
  const dsList = new DsList();
  dsList.append('a');
  dsList.append('b');
  dsList.append('c');
  assert(dsList.getItem(1) === 'b', 'Expected index 1 to be b');
}

function testSetItem() {
  const dsList = new DsList();
  dsList.append('a');
  dsList.append('b');
  dsList.append('c');
  dsList.setItem(1, 'd');
  assert(dsList.getItem(1) === 'd', 'Expected index 1 to be updated to d');
}

function testIter() {
  const dsList = new DsList();
  dsList.append('a');
  dsList.append('b');

  let count = 0;
  for (const _item of dsList) {
    count += 1;
  }

  assert(count === 2, 'Expected iteration to visit 2 items');
}

function testAdd() {
  const dsList1 = new DsList();
  dsList1.append('e');
  dsList1.append('b');

  const dsList2 = new DsList();
  dsList2.append('f');
  dsList2.append('f');

  const newDsList = dsList1.add(dsList2);
  assert(newDsList.toString() === "['e', 'b', 'f', 'f']", 'Expected concatenated list output');
}

function testInTrue() {
  const dsList = new DsList();
  dsList.append('e');
  dsList.append('b');
  dsList.append('f');
  dsList.append('f');

  const result = dsList.contains('b');
  assert(result === true, 'Expected contains to return true');
}

function testInFalse() {
  const dsList = new DsList();
  dsList.append('e');
  dsList.append('b');
  dsList.append('f');
  dsList.append('f');

  const result = dsList.contains('z');
  assert(dsList.length === 4 && result === false, 'Expected contains to return false');
}

function testClear() {
  const dsList = new DsList();
  dsList.append('e');
  dsList.append('b');
  dsList.append('f');
  dsList.append('f');

  dsList.clear();
  assert(dsList.length === 0, 'Expected length 0 after clear');
}

function checkFileForArrayUsage() {
  const filename = 'ListModule.js';
  const filepath = path.join(__dirname, filename);
  const content = fs.readFileSync(filepath, 'utf8');

  // Heuristic checks that avoid false positives for [Symbol.iterator]
  const bannedPatterns = [
    /\bnew\s+Array\b/, // new Array(...)
    /\bArray\.from\b/, // Array.from(...)
    /\bArray\s*\(/, // Array(...)
    /\.push\s*\(/,
    /\.pop\s*\(/,
    /\.shift\s*\(/,
    /\.unshift\s*\(/,
    /\.splice\s*\(/,
  ];

  for (const pattern of bannedPatterns) {
    if (pattern.test(content)) {
      return true;
    }
  }

  return false;
}

function testBuiltinListUsed() {
  const result = checkFileForArrayUsage();
  assert(result === false, 'Detected Array usage in ListModule.js');
}

function runTest(name, fn) {
  try {
    fn();
    successCount += 1;
  } catch (e) {
    failures.push([name, e && e.message ? e.message : String(e)]);
  }
}

function main() {
  runTest('testAppend', testAppend);
  runTest('testGetItem', testGetItem);
  runTest('testSetItem', testSetItem);
  runTest('testIter', testIter);
  runTest('testAdd', testAdd);
  runTest('testInTrue', testInTrue);
  runTest('testInFalse', testInFalse);
  runTest('testClear', testClear);
  runTest('testBuiltinListUsed', testBuiltinListUsed);

  console.log('');

  if (failures.length) {
    for (const [testName, message] of failures) {
      console.log(`ERROR: ${testName} -> ${message}`);
    }
    console.log(`\n${failures.length} tests failed`);
    console.log(`\n${successCount} tests passed`);
    console.log('');
    process.exitCode = 1;
    return;
  }

  console.log(`All ${successCount} tests passed`);
  console.log('');
}

main();
